### PR TITLE
HDFS: Allow overriding datanode registration addresses

### PR DESCRIPTION
**Pull request by nightkr**
# Description

Part of https://github.com/stackabletech/hdfs-operator/issues/174, implemented by rebuilding HDFS from source, after applying the required patches.

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
- [x] Changes are OpenShift compatible
- [x] All added packages (via microdnf or otherwise) have a comment on why they are added
- [x] Things not downloaded from Red Hat repositories should be mirrored in the Stackable repository and downloaded from there
- [x] All packages should have (if available) signatures/hashes verified
- [x] Does your change affect an SBOM? Make sure to update all SBOMs
```

### CODE REVIEW

File: hadoop/Dockerfile [hunk 1]
```diff
@@ -48,6 +36,40 @@ RUN curl --fail "https://repo.stackable.tech/repository/packages/jmx-exporter/jm
...
+    ./configure --prefix=/opt/protobuf && \
+    make "-j$(nproc)" && \
+    make install && \
+    cd /root && \
+    rm -rf /opt/protobuf-src
+
+ENV PROTOBUF_HOME /opt/protobuf
+ENV PATH "${PATH}:/opt/protobuf/bin"
+
+RUN curl --fail -L https://repo.stackable.tech/repository/packages/node/node-v${NODE}-linux-x64.tar.xz | tar -xJC . && \
```
Comment (nightkr) [inline hadoop/Dockerfile:57]: Required to build YARN. UBI8's nodejs is too old to be compatible.

Comment (nightkr) [reply hadoop/Dockerfile:57]: Note to self: move this into code comment

File: hadoop/Dockerfile [hunk 2]
```diff
@@ -48,6 +36,40 @@ RUN curl --fail "https://repo.stackable.tech/repository/packages/jmx-exporter/jm
...
+    ln -s /stackable/node-v${NODE}-linux-x64 /stackable/node
```
Comment (nightkr) [inline hadoop/Dockerfile:58]: Parametrize arch

File: hadoop/Dockerfile [hunk 3]
```diff
@@ -48,6 +36,40 @@ RUN curl --fail "https://repo.stackable.tech/repository/packages/jmx-exporter/jm
...
+ENV PATH "${PATH}:/stackable/node/bin"
+RUN corepack enable yarn && \
+    yarn config set ignore-engines true
+
+RUN cd /stackable/hadoop-${PRODUCT}-src && \
+    # Hadoop Pipes requires libtirpc to build, whose headers are not packaged in RedHat UBI
+    git apply < /stackable/patches/0001-disable-pipes.patch && \
+    # Datanode registration override is required for Listener Operator integration
+    # Developed at https://github.com/stackabletech/hadoop/tree/spike/override-datanode-id
+    git apply < /stackable/patches/0002-datanode-registration-override.patch && \
```
Comment (nightkr) [inline hadoop/Dockerfile:68]: Need a nicer way to maintain these patchsets... any ideas? @stackabletech/developers

Comment (lfrancke) [reply hadoop/Dockerfile:68]: I asked the same question a few weeks ago and ended up with
https://github.com/stackabletech/docker-images/tree/main/hbase/stackable/patches
https://github.com/stackabletech/docker-images/blob/main/hbase/Dockerfile#L41-L42

Comment (nightkr) [reply hadoop/Dockerfile:68]: I think that's fine for applying them, but editing .patches directly isn't exactly a super friendly workflow either :/

Comment (lfrancke) [reply hadoop/Dockerfile:68]: I've been looking for a solution for over 11 years
https://stackoverflow.com/questions/10705446/maintaining-multiple-patches-feature-branches-for-multiple-development-branches

Comment (siegfriedweber) [reply hadoop/Dockerfile]: The patching mechanism for the Hadoop image was implemented in #526. I took the liberty of merging the main branch into this one.

Comment (nightkr) [reply hadoop/Dockerfile]: I'll get back to merging that once the rest of the PR is finalized.

Comment (nightkr) [reply hadoop/Dockerfile]: Merged now.

**Comment by nightkr** [comment]
The 3.3.4 variant is tested against stackabletech/hdfs-operator@2cd93add0383fe55e9770c1cc0e50ae6126b2fae. The 3.3.6 variant is untested due to https://github.com/stackabletech/hdfs-operator/issues/440.

There is no 3.2.x backport, on the hypothesis that any new clusters should be running 3.3.x+, and any existing clusters have clearly already been fine without external access.. :P

**Review by siegfriedweber — DISMISSED** [review]
LGTM

You could add this to the changelog.

**Comment by nightkr** [comment]
Good catch, will do.

**Review by siegfriedweber — APPROVED** [review]

**Pull request merged**